source/Irrlicht/CGLTFMeshFileLoader.cpp

```diff
@@ -113,20 +115,134 @@ IAnimatedMesh* CGLTFMeshFileLoader::createMesh(io::IReadFile* file)
+		if (vertices->size() >= std::numeric_limits<u16>::max())
+			throw std::runtime_error("too many vertices");
+
+		// Apply the global transform along the parent chain.
+		for (auto &vertex : *vertices) {
+			parent->GlobalMatrix.transformVect(vertex.Pos);
+			// Apply scaling, rotation and translation (in that order) to the normal.
+			parent->GlobalMatrix.transformVect(vertex.Normal);
+			// Undo the translation, leaving us with scaling and rotation.
+			vertex.Normal -= parent->GlobalMatrix.getTranslation();
+			// Renormalize (length might have been affected by scaling).
+			vertex.Normal.normalize();
+		}
+		
+		auto maybeIndices = getIndices(meshIdx, j);
+		std::vector<u16> indices;
+		if (maybeIndices.has_value()) {
+			indices = std::move(maybeIndices.value());
+			for (u16 index : indices) {
+				if (index >= vertices->size())
+					throw std::runtime_error("index out of bounds");
+			}
+		} else {
+			// Non-indexed geometry
+			indices = std::vector<u16>(vertices->size());
+			for (std::size_t i = 0; i < vertices->size(); i += 3) {
+				// Reverse winding order per triangle
+				indices[i] = i + 2;
+				indices[i + 1] = i + 1;
+				indices[i + 2] = i;
+			}
+		}
+
+		auto *meshbuf = mesh->addMeshBuffer();
+		meshbuf->append(vertices->data(), vertices->size(),
+			indices.data(), indices.size());
+	}
+}
+
+// Base transformation between left & right handed coordinate systems.
+// This just inverts the Z axis.
+static core::matrix4 leftToRight = core::matrix4(
+	1, 0, 0, 0,
+	0, 1, 0, 0,
+	0, 0, -1, 0,
+	0, 0, 0, 1
+);
+static core::matrix4 rightToLeft = leftToRight;
+
+static core::matrix4 loadTransform(std::optional<std::variant<tiniergltf::Node::Matrix, tiniergltf::Node::TRS>> transform) {
+	if (!transform.has_value()) {
+		return core::matrix4();
+	}
+	core::matrix4 mat;
+	if (std::holds_alternative<tiniergltf::Node::Matrix>(*transform)) {
+		const auto &m = std::get<tiniergltf::Node::Matrix>(*transform);
+		// Note: Under the hood, this casts these doubles to floats.
+		mat = core::matrix4(
+			m[0], m[1], m[2], m[3],
+			m[4], m[5], m[6], m[7],
+			m[8], m[9], m[10], m[11],
+			m[12], m[13], m[14], m[15]);
+		// glTF uses column-major order, Irrlicht uses row-major order.
+		mat = mat.getTransposed();
+	} else {
+		const auto &trs = std::get<tiniergltf::Node::TRS>(*transform);
+		const auto &trans = trs.translation;
+		const auto &rot = trs.rotation;
+		const auto &scale = trs.scale;
+		core::matrix4 transMat;
+		transMat.setTranslation(core::vector3df(trans[0], trans[1], trans[2]));
+		core::matrix4 rotMat = core::quaternion(rot[0], rot[1], rot[2], rot[3]).getMatrix();
+		core::matrix4 scaleMat;
+		scaleMat.setScale(core::vector3df(scale[0], scale[1], scale[2]));
+		mat = transMat * rotMat * scaleMat;
+	}
+	return rightToLeft * mat * leftToRight;
+}
+
+void CGLTFMeshFileLoader::MeshExtractor::loadNode(
+		const std::size_t nodeIdx,
+		CSkinnedMesh* mesh,
+		CSkinnedMesh::SJoint *parent) const

```

Comment:
I'm seeing these two things get passed together quite a bit. Can the parent not be obtained from the mesh object? Otherwise maybe they could go together in a struct.

Reply:
> Can the parent not be obtained from the mesh object?

No, the parents vary per mesh. This is the parent of the node currently being processed.

I would suggest making the `CSkinnedMesh *mesh` a field of the `MeshExtractor` - then we can get rid of it in parameter lists of all `MeshExtractor` methods.

Reply:
Great idea, I like it.